Lib/ufo2ft/instructionCompiler.py

```diff
@@ -0,0 +1,194 @@
+# -*- coding: utf-8 -*-
+import logging
+
+from fontTools import ttLib
+from fontTools.pens.hashPointPen import HashPointPen
+from fontTools.ttLib.tables._g_l_y_f import (
+    OVERLAP_COMPOUND,
+    ROUND_XY_TO_GRID,
+    USE_MY_METRICS,
+)
+from ufo2ft.constants import (
+    TRUETYPE_INSTRUCTIONS_KEY,
+    TRUETYPE_ROUND_KEY,
+    TRUETYPE_METRICS_KEY,
+    TRUETYPE_OVERLAP_KEY,
+    OBJECT_LIBS_KEY,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class InstructionCompiler(object):
+    def __init__(self, ufo, ttf):
+        self.ufo = ufo
+        self.font = ttf
+
+    def _compile_program(self, key, table_tag):
+        assert table_tag in ("prep", "fpgm")
+        ttdata = self.ufo.lib.get(TRUETYPE_INSTRUCTIONS_KEY, None)
+        if ttdata:
+            formatVersion = ttdata.get("formatVersion", None)
+            if int(formatVersion) != 1:
+                logger.error(
+                    f"Unknown formatVersion {formatVersion} "
+                    f"in key '{key}', "
+                    f"table '{table_tag}' will be empty in font."
+                )
+                return
+            asm = ttdata.get(key, None)
+            if asm is not None:
+                self.font[table_tag] = table = ttLib.newTable(table_tag)
+                table.program = ttLib.tables.ttProgram.Program()
+                table.program.fromAssembly(asm)
+                # Roundtrip once, or if the font is dumped to XML before having
+                # been saved, the assembly code if will look awful.
+                table.program.fromBytecode(table.program.getBytecode())
+
+    def compile_fpgm(self):
+        self._compile_program("fontProgram", "fpgm")
+
+    def compile_glyf(self):
+        for name in sorted(self.ufo.keys()):
+            glyph = self.ufo[name]
+            ttdata = glyph.lib.get(TRUETYPE_INSTRUCTIONS_KEY, None)
+            if name not in self.font["glyf"]:
+                if ttdata is not None:
+                    logger.warning(
+                        f"Glyph '{name}' not found in font, "
+                        "skipping compilation of TrueType instructions "
+                        "for this glyph."
+                    )
+                logger.debug(f"UFO keys: {list(self.ufo.keys())}")
+                logger.debug(f"glyf keys: {list(self.font['glyf'].keys())}")
+                continue
+
+            glyf = self.font["glyf"][name]
```

Comment:
`glyf` is usually understood as the name of the whole table, not of an individual glyph within the glyf table. `glyf_glyph` or `ttglyph`